docs: cockroachdb storage backend configuration (#37880)
gravitational/teleport: docs/pages/reference/backends.mdx

teleport:
  data_dir: /var/lib/teleport_data
```

## CockroachDB

Teleport can use [CockroachDB](https://www.cockroachlabs.com/) as a storage backend

> to achieve high available and survive regional failures. You must take steps to 

Reviewer: ```suggestion
to achieve high availability and survive regional failures. You must take steps to 
```

protect access to CockroachDB in this configuration because that is where Teleport
secrets like keys and user records will be stored.

<Admonition
  type="warning"
  title="IMPORTANT"
>

>   CockroachDB can currently only be used to store Teleport's cluster state.

Reviewer: Starting to feel like we should have a matrix on this page that lists each backend and each use case (cluster state, audit log, session recordings, etc) to indicate which backends support which types of data.

required if the database already exists.

```
CREATE DATABASE database_name;
CREATE USER database_user;
GRANT CREATE ON DATABASE database_name TO database_user;
```

You must also enable change feeds in CockroachDB's cluster settings. Teleport
will configure this setting itself if granted `SYSTEM MODIFYCLUSTERSETTING`.

```
SET CLUSTER SETTING kv.rangefeed.enabled = true;
```

There are several ways to deploy and configure CockroachDB,
the details of which are not in scope for this guide. To learn about deploying
CockroachDB, see [CockroachDB's deployment options](https://www.cockroachlabs.com/docs/stable/choose-a-deployment-option).
To learn about how to configure multi-region survival goals, see
[multi-region survival goals](https://www.cockroachlabs.com/docs/stable/multiregion-survival-goals).

To configure Teleport to use CockroachDB as a storage backend:
- Configure all Teleport Auth Service instances to use the CockroachDB backend in the
  `storage` section of `teleport.yaml` as shown below.
- Deploy several Auth Service instances connected to the CockroachDB storage backend.
- Deploy several Proxy Service instances.

> - Make sure that the Proxy Service instances and all Teleport agent services that

Reviewer: This isn't specific to CockroachDB in any way, is it?

  connect directly to the the Auth Service have the `auth_server` configuration
  setting populated with the address of a load balancer for Auth Service
  instances.

```yaml
teleport:
  storage:
    type: cockroachdb

    # conn_string is a required parameter. It is a PostgreSQL connection string used

>     # to connect to `CockroachDB` using the PostgreSQL wire protocol. Client

Reviewer: ```suggestion
    # to connect to CockroachDB using the PostgreSQL wire protocol. Client
```

    # parameters may be specified using the URL. For a detailed list of available
    # parameters see https://www.cockroachlabs.com/docs/stable/connection-parameter
    #

>     # If your certificates are not stored at the default ~/.postgresql

Reviewer: Should we mention that we require mTLS? (Do we?)

    sync: NORMAL
    journal: WAL
```

## CockroachDB

Teleport can use [CockroachDB](https://www.cockroachlabs.com/) as a storage backend
to achieve high available and survive regional failures. You must take steps to 
protect access to CockroachDB in this configuration because that is where Teleport
secrets like keys and user records will be stored.

<Admonition
  type="warning"
  title="IMPORTANT"
>
  CockroachDB can currently only be used to store Teleport's cluster state.
  It cannot be used for Teleport's audit log in the same way that

>   [DynamoDB](#dynamodb) or Firestore](#firestore) wan.

Reviewer 2: ```suggestion
  [DynamoDB](#dynamodb) or [Firestore](#firestore) can.
```

    #
    # pool_max_conns is an additional parameter that determines the maximum
    # number of connections in the connection pool used for the cluster state
    # database (the change feed uses an additional connection), defaulting to
    # a value that depends on the number of available CPUs.
    conn_string: postgresql://user_name@database-address/teleport_backend?sslmode=verify-full&pool_max_conns=20

    # change_feed_conn_string is an optional parameter. When unspecified Teleport
    # will default to using the same value specified for conn_string. It may be used
    # to configure Teleport to use a different user or connection parameters when
    # establishing a change feed conneciton.
    #
    # If your certificates are not stored at the default ~/.postgresql
    # location, you will need to specify them with the sslcert, sslkey, and
    # sslrootcert parameters.
    change_feed_conn_string: postgresql://user_name@database-address/teleport_backend?sslmode=verify-full

    # ttl_job_cron is an optional parameter which configures the interval at which CockroachDB will expire backend
    # items based on their time to live. By default this is configured to run every

>     # 20 minutes. This is used by Teleport to cleanup old resources that are no longer

Reviewer 2: ```suggestion
    # 20 minutes. This is used by Teleport to clean up old resources that are no longer
```

    sync: NORMAL
    journal: WAL
```

## CockroachDB

Teleport can use [CockroachDB](https://www.cockroachlabs.com/) as a storage backend
to achieve high available and survive regional failures. You must take steps to 
protect access to CockroachDB in this configuration because that is where Teleport
secrets like keys and user records will be stored.

<Admonition
  type="warning"
  title="IMPORTANT"
>

>   `CockroachDB` can currently only be used to store Teleport's internal database in a
>   highly-available way. This will allow you to have multiple auth servers in your
>   cluster for a high availability deployment, but it will not also store Teleport audit events
>   for you in the same way that [DynamoDB](#dynamodb) or
>   [Firestore](#firestore) will.

Reviewer: This confused me on first read. It sounded like we were saying "be careful - you can only use this for HA and not for non-HA"
```suggestion
  CockroachDB can currently only be used to store Teleport's cluster state.
  It cannot be used for Teleport's audit log in the same way that
  [DynamoDB](#dynamodb) or Firestore](#firestore) wan.
```

I removed the backticks from CockroachDB because it is just a product name. It is not a command, code snippet, etc.

</Admonition>

At a minimum you must configure CockroachDB to allow Teleport to create tables.
Teleport will create the database if given permission to do so but this is not
required if the database already exists.

```
CREATE DATABASE database_name;
CREATE USER database_user;
GRANT CREATE ON DATABASE database_name TO database_user;
```

You must also enable change feeds in CockroachDB's cluster settings. Teleport
will configure this setting itself if granted `SYSTEM MODIFYCLUSTERSETTING`.

```
SET CLUSTER SETTING kv.rangefeed.enabled = true;
```

For more details on deploying CockroachDB see [CockroachDB's deployment options](https://www.cockroachlabs.com/docs/stable/choose-a-deployment-option).

> For more details on configuring multi-region survival goals in CockroachDB see

Reviewer: "For more details" suggests that there are at least some details here, but I don't see anything about configuring multi-region survival goals.

I would rephrase this. Perhaps something along the lines of this:

> There are several ways to configure multi-region deployments of CockroachDB,
> the details of which are not in scope for this guide. To learn about how to
> configure multi-region survival goals, see ...

Author: Makes sense I updated this for both the deployment and survival goals here https://github.com/gravitational/teleport/pull/37880/commits/b42399a329254ca6d30e7cbfdef2599b8ca8cb58
